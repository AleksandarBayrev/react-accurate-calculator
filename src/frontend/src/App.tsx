import React from 'react';
import './App.css';

export type AppProps = {}
export type AppState = {}

export class App extends React.Component<AppProps, AppState> {
  render() {
    return (
      <div className="App">
        <header className="App-header">
          <p>
            React Accurate Calculator
          </p>
        </header>
      </div>
    );
  }
}