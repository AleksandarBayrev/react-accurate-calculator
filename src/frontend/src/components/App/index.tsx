import React from 'react';
import '../../styles/App.css';
import { IApplicationConfiguration, ILogger, IMathWebApi, LogLevel } from '../../types';

export type AppProps = {
  configuration: IApplicationConfiguration
  mathWebApi: IMathWebApi
  logger: ILogger
}
export type AppState = {
  equation: string
  error: string
  result: string
}

export class App extends React.Component<AppProps, AppState> {
  private readonly configuration: IApplicationConfiguration;
  private readonly mathWebApi: IMathWebApi;
  private readonly logger: ILogger;

  constructor(props: AppProps) {
    super(props);
    this.configuration = props.configuration;
    this.mathWebApi = props.mathWebApi;
    this.logger = props.logger;
    this.state = {
      equation: '',
      error: '',
      result: ''
    };
  }

  private onChangeHandler = (e: any) => {
    this.setState({
      error: '',
      equation: e.target.value,
      result: ''
    });
  }

  private onClickHandler = async (e: any) => {
    if (!this.state.equation) {
      this.setState({
        error: 'Equation empty!'
      });
      return;
    }
    try {
      const result = await this.mathWebApi.calculate(this.state.equation);
      this.setState({
        result: result.resultAsString
      });
    } catch (err) {
      const error = `MathWebAPI: Error calculating equation for ${this.state.equation}`;
      this.setState({
        error
      });
      this.logger.Log(error, LogLevel.Error);
    }
  }

  private clearState = async () => {
    await new Promise((res, rej) => {
      setTimeout(() => {
        this.setState({
          error: '',
          result: '',
          equation: ''
        });
        res(this.state);
      });
    });
  }

  private renderResult() {
    const shouldRenderResult = this.state.equation.length && this.state.result.length;
    return (
      shouldRenderResult ?
      <div id="calculation-result">
        {this.state.equation}={this.state.result}
      </div>
      : <></>
    )
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <p>
            React Accurate Calculator
          </p>
          <p>
            <input type="text" onChange={this.onChangeHandler} placeholder="Input equation" value={this.state.equation} />
          </p>
          <p>
            <button onClick={this.onClickHandler}>Calculate</button>
          </p>
          <p>
            <button onClick={this.clearState}>Clear</button>
          </p>
          <p>
            {this.state.error.length ? this.state.error : this.renderResult()}
          </p>
        </header>
      </div>
    );
  }
}