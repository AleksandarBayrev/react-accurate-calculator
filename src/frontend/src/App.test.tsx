import renderer from 'react-test-renderer';
import { App } from './App';

describe('App UI Tests', () => {
  test('app snapshot', () => {
    const component = renderer.create(<App />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
