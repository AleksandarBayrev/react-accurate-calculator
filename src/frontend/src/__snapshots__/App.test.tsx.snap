// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App UI Tests app snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <p>
      React Accurate Calculator
    </p>
  </header>
</div>
`;
